pypuf/learner/evolution_strategies/becker.py

```diff
@@ -0,0 +1,311 @@
+import numpy as np
+import scipy.stats as stats
+import itertools
+from pypuf import tools
+from pypuf.simulation.arbiter_based.ltfarray import LTFArray
+
+class Reliability_based_CMA_ES():
+
+    # recommended properties of parameters:
+    #   pop_size=30
+    #   parent_size=10 (>= 0.3*pop_size)
+    #   priorities: linear low decreasing
+    def __init__(self, instance, pop_size, parent_size, priorities, challenge_num, repeat, unreliability):
+        self.transform = LTFArray.transform_id
+        self.combiner = LTFArray.combiner_xor
+        self.different_LTFs = np.zeros((instance.k, instance.n)) # learned LTFs
+        self.unreliability = unreliability  # proportion of unreliable challenges
+        self.epsilon = 0            # learned threshold for reliability
+        self.num_of_LTFs = 0        # number of different learned LTFs
+        self.new_LTF = np.empty(instance.n) # current learned LTF
+        self.terminate = False      # termination variable for learning 1 LTF
+        self.instance = instance    # simulation instance to be modelled
+        self.challenge_num = challenge_num    # number of challenges used
+        self.repeat = repeat        # frequency of same repeated challenge
+        self.individuals = np.zeros((parent_size, instance.n))  # weight_arrays
+        # mean, step_size,  pop_size,   parent_size,    priorities, cov_matrix,   path_cm,    path_ss
+        # m,    sigma,      lambda,     mu,             w_i,        C,            p_c,        p_sigma
+        self.mean = np.zeros(instance.n) # mean vector of distribution
+        self.step_size = 1              # distance to next distribution
+        self.pop_size = pop_size        # number of individuals per generation
+        self.parent_size = parent_size  # number of considered individuals
+        self.priorities = priorities    # array of consideration proportion
+        self.cov_matrix = np.identity(instance.n)  # shape of distribution
+        self.path_cm = np.matrix(np.zeros(instance.n)) # cumulated evolution path of covariance matrix
+        self.path_ss = np.zeros(instance.n) # cumulated evolution path of step size
+        # auxiliary constants
+        self.mu_w = 1 / np.sum(np.square(priorities))
+        self.c_mu = self.mu_w / instance.n**2
+        self.d_sigma = 1 + np.sqrt(self.mu_w / instance.n)
+        self.c_1 = 2 / instance.n**2
+        self.c_c = 4 / instance.n
+        self.c_sigma = 4 / instance.n
+        assert len(priorities) == parent_size
+        assert int(np.ndarray.sum(priorities)) == 1
+        assert (self.c_1 + self.c_mu <= 1)
+
+    def learn(self):
+        # this is the main learning method
+        # returns XOR-LTFArray with nearly same behavior as learned instance
+        self.epsilon = self.learn_epsilon(self.instance, self.unreliability)
+        while self.num_of_LTFs < self.instance.k:
+            self.new_LTF = self.learn_LTF()
+            if self.is_different_LTF():
```

Comment:
Refactor `self.new_LTF` into a method parameter for `self.is_different_LTF()`, it will make it easier to read: `if self.is_different_LTF(new_LTF):`, because it will immediately become clear what data `is_different_LTF` operates on.